### ISSUE TITLE
Cannot pass unhashable objects to xla.apply_primitive

### ISSUE THREAD

**Issue by dionhaefner** [google/jax#4280]
I define some custom XLA primitives and want to re-use the `xla_encode` implementation even in non-JIT contexts (so I don't need to maintain the same code twice). But I noticed that `xla.apply_primitive` doesn't work with unhashable types.

Test script:

```python
from jax import abstract_arrays
from jax.core import Primitive
from jax.interpreters import xla


p = Primitive("test_primitive")


def test_primitive(x, foo):
    return p.bind(x, foo=foo)


def test_primitive_impl(x, foo):
    return xla.apply_primitive(p, x, foo=foo)


def test_primitive_xla_encode(c, x, foo):
    raise NotImplementedError()


def test_primitive_abstract_eval(x, foo):
    return abstract_arrays.ShapedArray((1,), 'float')


p.def_impl(test_primitive_impl)
p.def_abstract_eval(test_primitive_abstract_eval)
xla.backend_specific_translations["cpu"][p] = test_primitive_xla_encode


if __name__ == '__main__':
    class SomeObject:
        __hash__ = None

    import jax
    import jax.numpy as jnp
    arr = jnp.ones(1)
    obj = SomeObject()

    try:
        jax.jit(lambda x: test_primitive(x, obj))(arr)
    except NotImplementedError:
        print('JIT works')

    try:
        test_primitive(arr, obj)
    except NotImplementedError:
        print('no-JIT works')
```

Output:

```bash
$ python test_apply_primitive.py
JIT works
Traceback (most recent call last):
  File "foo.py", line 47, in <module>
    test_primitive(arr, obj)
  File "foo.py", line 10, in test_primitive
    return p.bind(x, foo=foo)
  File "/Users/dion/.virtualenvs/science/lib/python3.7/site-packages/jax/core.py", line 274, in bind
    return self.impl(*args, **kwargs)
  File "foo.py", line 14, in test_primitive_impl
    return xla.apply_primitive(p, x, foo=foo)
  File "/Users/dion/.virtualenvs/science/lib/python3.7/site-packages/jax/interpreters/xla.py", line 224, in apply_primitive
    compiled_fun = xla_primitive_callable(prim, *unsafe_map(arg_spec, args), **params)
TypeError: unhashable type: 'SomeObject'
```

Why is hashability a constraint in this case when it seems to work just fine in the jitted code path?

**Comment by jekbradbury** [comment]
I see; you're trying to pass an unhashable object as a "param" (static, compile-time information) of a JAX primitive. This doesn't work because, outside a `jit`, JAX hashes the params of each primitive invoked for the per-primitive compilation cache. Perhaps we should allow non-hashable params to be compared by identity for the primitive compilation cache (although that seems like it might have been a bad idea in the jit cache), or perhaps you could implement a `__hash__` method that does that?

Basically, what's the situation where you'd need to "recompile" your primitive (re-do XLA translation and re-link/re-initialize your custom call)? Whenever `foo` changes its identity, or less often than that?

**Comment by dionhaefner** [comment]
Re-compilation on id change would be OK I think. The parameters should be static for a reason (one is comparable to a C enum, the other changes rarely). This is what happens in the JIT case, right? So consistent behavior sounds like a good idea to me, especially since "recompilation" should be a lot cheaper in the non-JIT case.

**Comment by dionhaefner** [comment]
FWIW, I solve this with a thin helper class for the moment:

```python
class Hashable:
    def __init__(self, obj):
        self.wrapped = obj

    def __hash__(self):
        return id(self.wrapped)

def make_hashable(obj):
    return Hashable(obj)

def unpack_hashable(obj):
    return obj.wrapped
```

and then use it like so:

```python
def test_primitive(x, foo):
    foo = make_hashable(foo)
    return p.bind(x, foo=foo)

def test_primitive_xla_encode(c, x, foo):
    foo = unpack_hashable(foo)
    # ...
```

The upside is that you can decide yourself what property you want to use for caching (`id` or something else).

**Comment by jekbradbury** [comment]
I agree that we want consistent behavior between the jit cache and the compiled primitive cache. But since we're actively considering switching the jit cache away from automatically assuming unhashable objects should be hashed by id, I kind of think forcing the user to define a `hash()` is reasonable here for now.

**Comment by hawkinsp** [comment]
Looking at this issue with 2023-eyes:

I think the hashable wrapper here is as it should be: it's up to the person defining primitives to ensure that their attributes are hashable, however they want to do that.

(We also no longer automatically wrap non-hashable values for `jit`, since that turned out to be something of a foot gun!)
